aries/aries_vcx_core/src/wallet/base_wallet/mod.rs

```diff
@@ -1,76 +1,63 @@
+use std::sync::Arc;
+
 use async_trait::async_trait;
-use public_key::Key;
-
-use super::record_tags::RecordTags;
-use crate::{
-    errors::error::VcxCoreResult,
-    wallet::{
-        base_wallet::{did_data::DidData, record::Record, search_filter::SearchFilter},
-        structs_io::UnpackMessageOutput,
-    },
+
+use self::{
+    did_wallet::DidWallet, issuer_config::IssuerConfig, record::AllRecords,
+    record_wallet::RecordWallet,
 };
+use crate::errors::error::VcxCoreResult;
 
 pub mod did_data;
+pub mod did_wallet;
+pub mod issuer_config;
+pub mod migrate;
 pub mod record;
+pub mod record_wallet;
 pub mod search_filter;
 
-pub trait BaseWallet: RecordWallet + DidWallet + Send + Sync + std::fmt::Debug {}
-
 #[async_trait]
-pub trait DidWallet {
-    async fn create_and_store_my_did(
-        &self,
-        seed: Option<&str>,
-        kdf_method_name: Option<&str>,
-    ) -> VcxCoreResult<DidData>;
+pub trait ImportWallet {
+    async fn import_wallet(&self) -> VcxCoreResult<()>;
+}
 
-    async fn key_for_did(&self, did: &str) -> VcxCoreResult<Key>;
+#[async_trait]
+pub trait ManageWallet {
+    async fn create_wallet(&self) -> VcxCoreResult<()>;
 
-    async fn replace_did_key_start(&self, did: &str, seed: Option<&str>) -> VcxCoreResult<Key>;
+    async fn open_wallet(&self) -> VcxCoreResult<Arc<dyn BaseWallet>>;
 
-    async fn replace_did_key_apply(&self, did: &str) -> VcxCoreResult<()>;
+    async fn delete_wallet(&self) -> VcxCoreResult<()>;
+}
 
-    async fn sign(&self, key: &Key, msg: &[u8]) -> VcxCoreResult<Vec<u8>>;
+#[async_trait]
+pub trait BaseWallet: RecordWallet + DidWallet + Send + Sync + std::fmt::Debug {
+    async fn export_wallet(&self, path: &str, backup_key: &str) -> VcxCoreResult<()>;
 
-    async fn verify(&self, key: &Key, msg: &[u8], signature: &[u8]) -> VcxCoreResult<bool>;
+    async fn close_wallet(&self) -> VcxCoreResult<()>;
 
-    async fn pack_message(
-        &self,
-        sender_vk: Option<Key>,
-        receiver_keys: Vec<Key>,
-        msg: &[u8],
-    ) -> VcxCoreResult<Vec<u8>>;
+    async fn configure_issuer(&self, key_seed: &str) -> VcxCoreResult<IssuerConfig>;
 
-    async fn unpack_message(&self, msg: &[u8]) -> VcxCoreResult<UnpackMessageOutput>;
+    async fn all(&self) -> VcxCoreResult<Box<dyn AllRecords + Send>>;
```

Comment:
Why custom trait instead of simply implementing an iterator? Why on `BaseWallet` instead of `RecordWallet`?

Reply:
Because I need a custom abstraction for how different wallet implementations fetch all records. vdrtools wallet provides [WalletIterator](https://github.com/hyperledger/aries-vcx/blob/main/aries/misc/legacy/libvdrtools/indy-wallet/src/iterator.rs#L14-L38) which does not conform to `std::iter::Iterator` nor to `std::async_iter::AsyncIterator`. 

Moved to `RecordWallet`.

Reply:
I asked why it's implemented on `BaseWallet` instead of `RecordWallet`, I am not arguing for moving it (or not).

Also, I am sorry, but I don't see how "I need a custom abstraction for how different wallet implementations fetch all records" is explanation for creating a custom trait instead of implementing an iterator.

Reply:
`RecordWallet` operates only on records, so it was implemented in `BaseWallet` because it fetches both records and keys. Even though keys are "key records" since everything is stored in the same table. Indy wallet fetches everything with a single method while askar has separate methods for keys and records. 

Instead of just implementing an iterator, a wallet specific trait was created because 
[getting all records](https://github.com/hyperledger/aries-vcx/blob/main/aries/misc/legacy/libvdrtools/indy-wallet/src/wallet.rs#L472) for indy gives me [WalletIterator](https://github.com/hyperledger/aries-vcx/blob/main/aries/misc/legacy/libvdrtools/indy-wallet/src/iterator.rs#L14-L38) which is async and I do not see a way how to implement a sync trait using async function.

I could implement async iterator but that requires bringing in additional dependency.